File: isotonic.r
isotonic.maxflow <- function (C) 
{
    n <- nrow(C)
    F <- matrix(nrow = n, ncol = n, 0)
    ready <- FALSE
    labels <- matrix(nrow = n, ncol = 2, 0)
    while (!ready) {
        labels[, ] <- 0
        labels[1, ] <- c(1, 10000)
        queue <- c(1)
        index <- 1
        while (labels[n, 1] == 0) {
            if (length(queue) == 0) {
                return(labels)
            }
            v <- queue[index]
            queue <- queue[-index]
            index1 <- c(1:n)[C[v, ] > 0 & labels[, 1] == 0]
            index2 <- c(1:n)[C[, v] > 0 & labels[, 1] == 0]
            for (i in index1) {
                if (F[v, i] < C[v, i]) {
                  labels[i, 1] <- v
                  labels[i, 2] <- min(labels[v, 2], C[v, i] - 
                    F[v, i])
                  queue <- c(queue, i)
                }
            }
            for (i in index2) {
                if (F[i, v] > 0) {
                  labels[i, 1] <- v
                  labels[i, 2] <- min(labels[v, 2], F[i, v])
                  queue <- c(queue, i)
                }
            }
        }
        v <- n
        lambda <- labels[n, 2]
        while (v != 1) {
            u <- labels[v, 1]
            if (C[u, v] > 0) {
                F[u, v] <- F[u, v] + lambda
            }
            else {
                F[v, u] <- F[v, u] - lambda
            }
            v <- u
        }
    }
}


isotonic.monoreg <- function (ordermat, g, w, subset) 
# implements the algorithm of 1. Maxwell and Muckstadt, 2. Spouge, Wan and Wilbur, 3. Picard
# ordermat is an incidence matrix representing the partial order
# that is: ordermat[i,j] == 1 iff i <= j.
# g contains the unconstrained estimates to be made monotone
# w[i] is the weight of g[i], typically the number of observations on which the unconstrained
# estimate g[i] has been computed
{
    m <- length(subset)
    # trivial case: the subset only contains one element
    if (m == 1) {
        g.star.index <- subset
        g.star.value <- g[subset]
        blocks <- list(list(block = subset, av = g[subset]))
    }
    else {
        # compute the weighted average of g, with weights w, on the subset
        av <- sum(g[subset] * w[subset])/sum(w[subset])
        print(av)
        b <- w[subset] * (g[subset] - av)
        index.pos <- c(1:m)[b > 0]
        index.neg <- c(1:m)[b < 0]

        if (length(index.pos) == 0 && length(index.neg) == 0) {
            g.star.index <- subset
            g.star.value <- rep(av, m)
            blocks <- list(list(block = subset, av = av))
        }
        else {
            network <- matrix(nrow = m + 2, ncol = m + 2, 0)
            # add directed edge with "infinite" capacity from x to y iff x <= y.
            network[2:(m + 1), 2:(m + 1)] <- ordermat[subset,subset] * 1e+05
            
            # add edge from source to elements with b > 0 with capacity b
            network[1, index.pos + 1] <- b[index.pos]
            # add edge from elements with b < 0 to sink, with capacity -b
            network[index.neg + 1, m + 2] <- -b[index.neg]
            # solve the maximum flow problem on the constructed network to find the maximum upper set
            labels <- isotonic.maxflow(network)
            #print(sum(labels))
            temp <- as.vector(labels[2:(m + 1), 2])
            # determine "projection pair": the maximum upper set and minimum lower set
            index <- c(1:m)[temp > 0]
            upset <- subset[index]
            lowset <- subset[-index]
            
            if (length(upset) == 0) {
                g.star.index <- subset
                g.star.value <- rep(av, m)
                blocks <- list(list(block = subset, av = av))
            }
            else if (length(lowset) == 0) {
                g.star.index <- subset
                g.star.value <- rep(av, m)
                blocks <- list(list(block = subset, av = av))
            }
            else {
                # recurse on the maximum upperset U and minimum lower set L
                up <- isotonic.monoreg(ordermat, g, w, upset)
                low <- isotonic.monoreg(ordermat, g, w, lowset)
                # g*|U = (g|U)* and g*|L = (g|L)*
                g.star.index <- c(low$g.star.index, up$g.star.index)
                g.star.value <- c(low$g.star.value, up$g.star.value)
                blocks <- c(low$blocks, up$blocks)
            }
        }
    }
    list(g.star.index = g.star.index, g.star.value = g.star.value, 
        blocks = blocks)
}


isotonic.main <- function (ordermat, y, w = rep(1, length(y))) 
{
    n <- length(y)
    out <- isotonic.monoreg(ordermat, y, w, 1:n)
    out$g.star.value[order(out$g.star.index)]
}

#isotonic.main(ordermat.quasi2, c(0.4,0.2,0.6,0.4))

#library(reticulate)
#pd <- import("pandas")
ordermat<- pd$read_pickle("ordermat.pickle")
#pima = read.csv(file = 'pima.csv', header = F)
#y = pima$V9
res = isotonic.main(ordermat, y)
#sum = 0
#res
##for (g in res) {
#  sum = sum + min(1-g, g)
#}
#sum

File: isotonic.py
import numpy as np
import pandas as pd

UPPERBOUND = 1e+15

class Isotonic:
    def __init__(self, ordermat: np.array, y: np.array, w: np.array=None) -> None:
        if w is None:
            w = np.repeat(1, len(y))
        n = len(y)
        y = np.pad(y, (1,0), 'constant', constant_values=-1)
        w = np.pad(w, (1,0), 'constant', constant_values=-1)
        ordermat = np.pad(ordermat, ((1,0),(1,0)), 'constant', constant_values=-1)
        out = self.monoreg(ordermat, y, w, np.arange(1,n+1))
        indices = np.argsort(out['g_star_index'])
        self.result = out['g_star_value'][indices]
    
    def maxflow(self, C):
        """
        Maxflow on the network C, returns labels of nodes that belong to maximum upperset
        """
        n = len(C) - 1
        F = np.zeros((n,n)) 
        F = np.pad(F, ((1,0), (1,0)), 'constant', constant_values=-1)
        ready = False
        labels = np.zeros((n,2))
        labels = np.pad(labels, ((1,0), (1,0)), 'constant', constant_values=-1)
        while not ready:
            labels.fill(0)
            labels[1, 1:] = np.array([1, 1000000000000000]) #TODO: vraag Ad wat deze grote constant is? kan die worden vervangen door upperbound?
            queue = np.array([0,1])
            index = 1
            while labels[n, 1] == 0:
                if len(queue) - 1 == 0:
                    return labels
                v = queue[index]
                queue = np.delete(queue, index, axis=0)
                index1 = np.arange(1,n+1)[ (C[v, 1:] > 0) & (labels[1:, 1] == 0) ]
                index2 = np.arange(1,n+1)[ (C[1:, v] > 0) & (labels[1:, 1] == 0) ]
                for i in index1:
                    if F[v, i] < C[v, i]:
                        labels[i, 1] = v
                        labels[i, 2] = min(labels[v, 2], C[v, i] - F[v, i])
                        queue = np.concatenate([queue,[i]])

                for i in index2:
                    if F[i, v] > 0:
                        labels[i, 1] = v
                        labels[i, 2] = min(labels[v, 2], F[i, v])
                        queue = np.concatenate([queue,[i]])
            v = n
            lambda_ = labels[n, 2]
            while v != 1:
                u = int(labels[v,1])
                if C[u, v] > 0:
                    F[u, v] = F[u, v] + lambda_
                else:
                    F[v, u] = F[v, u] - lambda_
                v = u
        return labels
    
    def monoreg(self, ordermat: np.array, g: np.array, w: np.array, subset: np.array) -> pd.DataFrame:
        """
        Implements the algorithm of 1. Maxwell and Muckstadt, 2. Spouge, Wan and Wilbur, 3. Picard
        ordermat is an incidence matrix representing the partial order, that is: ordermat[i,j] == 1 iff i <= j. 
        g contains the unconstrained estimates to be made monotone
        w[i] is the weight of g[i], typically the number of observations on which the unconstrained
        estimate g[i] has been computed
        """
        m = len(subset)
        # Trivial case: the subset only contains one element
        if m == 1:
            g_star_index = subset
            g_star_value = g[subset]
        else:
            # Compute the weighted average of g, with weights w, on the subset
            av = np.sum(g[subset] * w[subset]) / np.sum(w[subset])
            b = w[subset] * (g[subset] - av)
            b = np.pad(b, (1,0), 'constant', constant_values=-1)
            index_pos = np.arange(1, m + 1)[b[1:]> 0]
            index_neg = np.arange(1, m + 1)[b[1:] < 0]

            if (len(index_pos) == 0 and len(index_neg) == 0):
                g_star_index = subset
                g_star_value = np.repeat(av, m)
            else:
                # Include source and sink
                network = np.zeros((m+2, m+2)) 
                network = np.pad(network, ((1,0),(1,0)), 'constant', constant_values=None)

                # Add directed edge with "infinite" capacity from x to y iff x <= y.
                network[2:(m + 2), 2:(m + 2)] = ordermat[subset, :][:, subset] * UPPERBOUND

                # Add edge from source to elements with b > 0 with capacity b
                network[1, index_pos + 1] = b[index_pos]
                # Add edge from elements with b < 0 to sink, with capacity -b
                network[index_neg + 1, m + 2] = -b[index_neg]
                # Solve the maximum flow problem on the constructed network to find the maximum upper set
                labels = self.maxflow(network)
                temp = np.reshape(labels[2:(m+2), 2], -1, 'F')
                index = np.arange(1,m+1)[temp > 0]
                subset_padded = np.pad(subset, (1,0), constant_values=-1)
                upset = subset_padded[index]
                lowset = np.delete(subset_padded, np.append(index, 0))

                if (len(upset) == 0):
                    g_star_index = subset
                    g_star_value = np.repeat(av, m)
                
                elif (len(lowset) == 0):
                    g_star_index = subset
                    g_star_value = np.repeat(av, m)
                
                else:
                    # Recurse on the maximum upperset U and minimum lower set L
                    up = self.monoreg(ordermat, g, w, upset)
                    low = self.monoreg(ordermat, g, w, lowset)
                    # Thus we proceed with: g*|U = (g|U)* and g*|L = (g|L)*
                    g_star_index = np.concatenate((np.array(low['g_star_index']), np.array(up['g_star_index'])))
                    g_star_value = np.concatenate((np.array(low['g_star_value']), np.array(up['g_star_value'])))
        result = pd.DataFrame({'g_star_index': g_star_index, 'g_star_value': g_star_value})
        return result

# Isotonic(np.array([[0,1,1,1], [0,0,0,1], [0,0,0,1], [0,0,0,0]]), np.array([0.4, 0.2, 0.6, 0.4]))
    